main/src/ca/uwaterloo/flix/language/phase/EnumGen.scala

```diff
@@ -0,0 +1,650 @@
+      case Type.Apply(Type.Enum(s, _), _) => s
+      case Type.Enum(s, _) => s
+      case _ => throw InternalCompilerException(s"Unexpected type: `$tpe'.")
+    }}
+
+    // 3. Generate byteCodes of Enum Interface.
+    val enumInterfaces : Map[EnumSym, (QualName, Array[Byte])] = enumsGroupedBySymbol.map{ case (sym, _) =>
+      val name = CodegenHelper.getEnumInterfaceName(sym)
+      sym -> (name, compileEnumInterface(name)(flix))
+    }.toMap // Despite IDE highlighting, this is actually necessary.
+
+    // 4. Separate primitive enums and object enums.
+    val enumsGroupedBySymbolAndFieldType : Map[EnumSym, (Set[(String, Type)], Set[String])] = enumsGroupedBySymbol.map{case (sym, li) =>
+      val (primEnums, enumsWithObjField) = li.partition(_._2._2.isPrimitive)
+      sym -> (primEnums.map(x => (x._2._1, x._2._2)).toSet, enumsWithObjField.map(x => x._2._1).toSet)
+    }.toMap // Despite IDE highlighting, this is actually necessary.
+
+    // 5. Generate bytecode for each enum case.
+    val enumByteCodes = enumsGroupedBySymbolAndFieldType.map{ case (sym, (primEnums, objectEnums)) =>
+      val generatedPrimEnums : Map[(String, Type), Array[Byte]] = primEnums.map{ case (name, tpe) =>
+        val qualName = CodegenHelper.getEnumCaseName(sym, name, Some(tpe))
+        val interface = enumInterfaces(sym)._1
+        (name, tpe) -> EnumGen.compileEnumClass(qualName, interface, Some(tpe), interfaceNames, flix.javaVersion)
+      }.toMap
+      val generatedObjectEnums : Map[String, Array[Byte]] = objectEnums.map{ name =>
+        val fullName = CodegenHelper.getEnumCaseName(sym, name, None)
+        val interface = enumInterfaces(sym)._1
+        name -> EnumGen.compileEnumClass(fullName, interface, None, interfaceNames, flix.javaVersion)
+      }.toMap
+      sym -> (generatedPrimEnums, generatedObjectEnums)
+    }
+
+    val e = System.nanoTime() - t
+    root.copy(time = root.time.copy(enumGen = e), enumInterfaceByteCodes = enumInterfaces, enumClassByteCodes = enumByteCodes).toSuccess
+  }
+
+  /**
+    * Generates an interface for each enum.
+    * Each case of the enum implements this interface. This interface extends `TagInterface`.
+    * For example, for enum `Result` we create:
+    *
+    * package ca.waterloo.flix.enums.Result;
+    * import ca.uwaterloo.flix.api.TagInterface;
+    * public interface EnumInterface extends TagInterface {
+    * }
+    *
+    * @param qualName Qualified Name of the interface to be generated
+    * @return byte code representation of the class
+    */
+  def compileEnumInterface(qualName: QualName)(flix: Flix): Array[Byte] = {
+    val visitor = new ClassWriter(0)
+
+    visitor.visit(flix.javaVersion, ACC_PUBLIC + ACC_ABSTRACT + ACC_INTERFACE, decorate(qualName), null,
+      asm.Type.getInternalName(Constants.objectClass), Array(asm.Type.getInternalName(Constants.tagInterface)))
```

Comment:
Could we extract the last two arguments into some local variables with informative names? I am thinking one of these is the superclass.